go/teams/service_helper.go

```diff
@@ -436,61 +436,98 @@ type AddMembersRes struct {
-					return err
-				}
-				return NewAddMembersError(user.AssertionOrEmail, err)
+				return err
 			}
-			var normalizedUsername libkb.NormalizedUsername
-			if !username.IsNil() {
-				normalizedUsername = username
+
+			tx := CreateAddMemberTx(team)
+			type sweepEntry struct {
+				Assertion string
+				UV        keybase1.UserVersion
 			}
-			res[i] = AddMembersRes{
-				Invite:   invite,
-				Username: normalizedUsername,
+			var sweep []sweepEntry
+			for i, user := range users {
+				upak, single, doInvite, err := tx.ResolveUPKV2FromAssertionOrEmail(mctx, user.AssertionOrEmail)
+				if err != nil {
+					return NewAddMembersError(user.AssertionOrEmail, err)
+				}
+
+				if _, ok := restrictedUsers[upak.Uid]; ok {
+					// skip users with contact setting restrictions
+					continue
+				}
+
+				username, uv, invite, err := tx.AddOrInviteMemberByUPKV2(ctx, upak, single, doInvite, user.AssertionOrEmail, user.Role, user.BotSettings)
+				if err != nil {
+					if _, ok := err.(AttemptedInviteSocialOwnerError); ok {
+						return err
+					}
+					return NewAddMembersError(user.AssertionOrEmail, err)
+				}
+				var normalizedUsername libkb.NormalizedUsername
+				if !username.IsNil() {
+					normalizedUsername = username
+				}
+
+				res[i] = AddMembersRes{
+					Invite:   invite,
+					Username: normalizedUsername,
+				}
+				if !uv.IsNil() {
+					sweep = append(sweep, sweepEntry{
+						Assertion: user.AssertionOrEmail,
+						UV:        uv,
+					})
+				}
 			}
-			if !uv.IsNil() {
-				sweep = append(sweep, sweepEntry{
-					Assertion: user.AssertionOrEmail,
-					UV:        uv,
-				})
+			// Try to mark completed any invites for the users' social assertions.
+			// This can be a time-intensive process since it involves checking proofs.
+			// It is limited to a few seconds and failure is non-fatal.
+			timeoutCtx, timeoutCancel := context.WithTimeout(ctx, 2*time.Second)
+			for _, x := range sweep {
+				if err := tx.CompleteSocialInvitesFor(timeoutCtx, x.UV, x.Assertion); err != nil {
+					g.Log.CWarningf(ctx, "Failed in CompleteSocialInvitesFor(%v, %v) -> %v", x.UV, x.Assertion, err)
+				}
 			}
+			timeoutCancel()
+
+			return tx.Post(libkb.NewMetaContext(ctx, g))
 		}
+	}
 
-		// Try to mark completed any invites for the users' social assertions.
-		// This can be a time-intensive process since it involves checking proofs.
-		// It is limited to a few seconds and failure is non-fatal.
-		timeoutCtx, timeoutCancel := context.WithTimeout(ctx, 2*time.Second)
-		for _, x := range sweep {
-			if err := tx.CompleteSocialInvitesFor(timeoutCtx, x.UV, x.Assertion); err != nil {
-				g.Log.CWarningf(ctx, "Failed in CompleteSocialInvitesFor(%v, %v) -> %v", x.UV, x.Assertion, err)
+	const nRetries = 2
+	restrictedUsers := make(map[keybase1.UID]bool)
+	for i := 0; i < nRetries; i++ {
+		mctx.Debug("| Retry AddMembers(%v)", i)
+		err = RetryIfPossible(ctx, g, retryableFunc(restrictedUsers))
+		if isTeamContactSettingsBlock(err) {
```

Comment:
don't do this, there is already an error with these fields parsed out (`libkb.TeamContactSettingsBlockedError`)

Reply:
stylistically, this might be cleaner like this:

```go
try := func() { /* try to add */ }
err := try()
if err == theRightTypeOfError {
   /* change something */
   try()
}
```

Reply:
the err here is of type `libkb.AppStatusError`. 
i see the logic in UnwrapError to return a `libkb.TeamContactSettingsBlockedError` https://github.com/keybase/client/blob/e2ac1fa45440f7f9315896710232d08a225d9ac8/go/libkb/rpc_exim.go#L222 but AddMembers (as called in my tests) aren't reaching it... am i looking in the wrong place?

Reply:
Right, then there might be a bug.... I wonder what it is